### PR TITLE
Support for changing output options at run-time

### PR DESCRIPTION
**Pull request by Wengier**
This adds support for changing options like output and transparency from the DOS command shell at run-time. For example, the command ``output texture`` changes the current output to "texture" from another one, and ``transparency 30`` changes the transparency to 30. Allowing run-time change of output options is also helpful for others features like Voodoo and TTF.

P.S. The bulk of code is moved from GUI_StartUp to set_output function to make dynamic change possible.

### CODE REVIEW

File: include/video.h [hunk 1]
```diff
@@ -74,6 +74,10 @@ void GFX_EndUpdate( const Bit16u *changedLines );
 void GFX_GetSize(int &width, int &height, bool &fullscreen);
 void GFX_LosingFocus(void);
 
+bool wants_stretched_pixels();
+void remove_window();
+void set_output(Section *sec, bool should_stretch_pixels);
+
```
Comment (kcgen) [inline include/video.h:80]: Suggest using a single prototype, `void GFX_RegenerateWindow()` (more below)

File: src/gui/sdl_mapper.cpp [hunk 2]
```diff
@@ -3002,6 +3003,13 @@ void MAPPER_BindKeys(Section *sec)
 
 	if (SDL_GetModState()&KMOD_NUM)
 		MAPPER_TriggerEvent(num_lock_event, false);
+
+	if (!first_window) {
+		remove_window();
+		set_output(sec, wants_stretched_pixels());
+		GFX_ResetScreen();
+	}
+	first_window = false;
```
Comment (kcgen) [inline src/gui/sdl_mapper.cpp:3012]: Suggest moving this functionality into `sdlmain.cpp` packaged as a function, something like:

``` c++
void GFX_RegenerateWindow() {
	if (!first_window) {
		remove_window();
		set_output(sec, wants_stretched_pixels());
		GFX_ResetScreen();
	}
	first_window = false;
}
```

(Made available to the mapper with the prototype in `video.h`)

File: src/gui/sdlmain.cpp [hunk 3]
```diff
@@ -849,10 +849,7 @@ static SDL_Window *SetWindowMode(SCREEN_TYPES screen_type,
 
 		last_type = screen_type;
 
-		if (sdl.window) {
-			SDL_DestroyWindow(sdl.window);
-			sdl.window = nullptr;
-		}
+		remove_window();
```
Comment (kcgen) [inline src/gui/sdlmain.cpp:853]: Nice DRY cleanup @Wengier !
:+1:

**Review by kcgen — CHANGES_REQUESTED** [review]
Just a small suggestion.

File: include/video.h (hunk 1 above)
Comment (Wengier) [reply include/video.h]: Done.

File: src/gui/sdl_mapper.cpp (hunk 2 above)
Comment (Wengier) [reply src/gui/sdl_mapper.cpp]: Done.

**Comment by Wengier** [comment]
@kcgen Thanks for the review. I have addressed them accordingly.

File: include/video.h [hunk 4]
```diff
@@ -72,7 +72,12 @@ void GFX_SwitchFullScreen(void);
 bool GFX_StartUpdate(uint8_t * &pixels, int &pitch);
 void GFX_EndUpdate( const Bit16u *changedLines );
 void GFX_GetSize(int &width, int &height, bool &fullscreen);
-void GFX_LosingFocus(void);
+void GFX_LosingFocus();
+void GFX_RegenerateWindow(Section *sec);
+
+bool wants_stretched_pixels();
+void remove_window();
+void set_output(Section *sec, bool should_stretch_pixels);
```
Comment (kcgen) [inline include/video.h:80]: Thanks @Wengier. 
Hopefully these three prototypes (wants_stretched_pixels, remove_window, set_output) can now be dropped from the `video.h` header?

File: src/gui/sdlmain.cpp [hunk 5]
```diff
@@ -2698,6 +2695,207 @@ static SDL_Rect calc_viewport_pp(int win_width, int win_height)
 		return calc_viewport_fit(width, height);
 }
 
+void remove_window()
+{
+	if (sdl.window) {
+		SDL_DestroyWindow(sdl.window);
+		sdl.window = nullptr;
+	}
+}
```
Comment (kcgen) [inline src/gui/sdlmain.cpp:2704]: `remove_window` is only used locally in this cpp file, so should be `static` function.
Might have to move this function up to line ~832 so it's visible before its used.

File: src/gui/sdlmain.cpp [hunk 6]
```diff
@@ -2281,7 +2278,7 @@ static bool detect_resizable_window()
 #endif // C_OPENGL
 }
 
-static bool wants_stretched_pixels()
+bool wants_stretched_pixels()
```
Comment (kcgen) [inline src/gui/sdlmain.cpp:2281]: Hopefully can keep this static too.

File: src/gui/sdlmain.cpp [hunk 7]
```diff
@@ -2698,6 +2695,207 @@ static SDL_Rect calc_viewport_pp(int win_width, int win_height)
...
 
+void remove_window()
+{
+	if (sdl.window) {
+		SDL_DestroyWindow(sdl.window);
+		sdl.window = nullptr;
+	}
+}
+
+void set_output(Section *sec, bool should_stretch_pixels)
```
Comment (kcgen) [inline src/gui/sdlmain.cpp:2706]: `static` as well.

File: src/gui/sdlmain.cpp [hunk 8]
```diff
@@ -3094,10 +3144,21 @@ void GFX_LosingFocus()
...
 
+void GFX_RegenerateWindow(Section *sec) {
+	if (!first_window) {
+		const auto section = static_cast<const Section_prop *>(sec);
+		if (strcmp(section->Get_string("output"), "surface"))
+			remove_window();
+		set_output(sec, wants_stretched_pixels());
+		GFX_ResetScreen();
+	}
+	first_window = false;
```
Comment (kcgen) [inline src/gui/sdlmain.cpp:3159]: (Very minor/optional suggestion to drop a nesting level and avoid redundant assignment):
``` c++
	if (first_window) {
		first_window = false;
		return;
	}
	const auto section = static_cast<const Section_prop *>(sec);
	if (strcmp(section->Get_string("output"), "surface"))
		remove_window();
	set_output(sec, wants_stretched_pixels());
	GFX_ResetScreen();
```

File: include/video.h (hunk 4 above)
Comment (Wengier) [reply include/video.h]: Removed them from the header. Done.

File: src/gui/sdlmain.cpp (hunk 5 above)
Comment (Wengier) [reply src/gui/sdlmain.cpp]: Done.

File: src/gui/sdlmain.cpp (hunk 6 above)
Comment (Wengier) [reply src/gui/sdlmain.cpp]: Done.

File: src/gui/sdlmain.cpp (hunk 7 above)
Comment (Wengier) [reply src/gui/sdlmain.cpp]: Done.

File: src/gui/sdlmain.cpp (hunk 8 above)
Comment (Wengier) [reply src/gui/sdlmain.cpp]: Done accordingly.

**Review by kcgen — APPROVED** [review]
All looks good, thanks @Wengier !

**Pull request merged**